src/chainparams.cpp

```diff
@@ -466,14 +474,60 @@ class CRegTestParams : public CChainParams {
     {
-        consensus.vDeployments[d].nStartTime = nStartTime;
-        consensus.vDeployments[d].nTimeout = nTimeout;
+        consensus.vDeployments[d].startheight = startheight;
+        consensus.vDeployments[d].timeoutheight = timeoutheight;
+        consensus.vDeployments[d].m_min_activation_height = min_activation_height;
     }
     void UpdateActivationParametersFromArgs(const ArgsManager& args);
 };
 
+bool CChainParams::CheckVBitsHeights(std::string& error, int startheight, int timeoutheight, int min_activation_height) const
+{
+    // Special always or never active cases
+    if ((startheight == Consensus::BIP9Deployment::NEVER_ACTIVE && timeoutheight == Consensus::BIP9Deployment::NEVER_ACTIVE)
+        || startheight == Consensus::BIP9Deployment::ALWAYS_ACTIVE) {
+        return true;
+    }
+    if ((startheight == Consensus::BIP9Deployment::NEVER_ACTIVE && timeoutheight != Consensus::BIP9Deployment::NEVER_ACTIVE)
+        || (startheight != Consensus::BIP9Deployment::NEVER_ACTIVE && timeoutheight == Consensus::BIP9Deployment::NEVER_ACTIVE)) {
+        error = strprintf("When one of startheight or timeoutheight is %d, both must be %d", Consensus::BIP9Deployment::NEVER_ACTIVE, Consensus::BIP9Deployment::NEVER_ACTIVE);
+        return false;
+    }
+
+    // Actual params must be on retarget block
+    if (startheight < 0) {
+        error = strprintf("Invalid startheight (%d), cannot be negative (except for never or always active special cases)", startheight);
+        return false;
+    }
+    if (timeoutheight < 0) {
+        error = strprintf("Invalid timeoutheight (%d), cannot be negative (except for never or always active special cases)", timeoutheight);
+        return false;
+    }
+    if (startheight % consensus.nMinerConfirmationWindow != 0) {
+        error = strprintf("Invalid startheight (%d), must be a multiple of %d", startheight, consensus.nMinerConfirmationWindow);
+        return false;
+    }
+    if (timeoutheight != Consensus::BIP9Deployment::NO_TIMEOUT && timeoutheight % consensus.nMinerConfirmationWindow != 0) {
+        error = strprintf("Invalid timeoutheight (%d), must be a multiple of %d", timeoutheight, consensus.nMinerConfirmationWindow);
+        return false;
+    }
+    if (min_activation_height < 0) {
+        error = strprintf("Invalid minimum activation height (%d), cannot be negative", min_activation_height);
+        return false;
+    }
+    if (min_activation_height % consensus.nMinerConfirmationWindow != 0) {
+        error = strprintf("Invalid minimum activation height (%d), must be a multiple of %d", min_activation_height, consensus.nMinerConfirmationWindow);
+        return false;
+    }
+    if (timeoutheight < startheight + (int)consensus.nMinerConfirmationWindow) {
+        error = strprintf("Invalid timeoutheight (%d), must be at least one period greater than the startheight (%d)", timeoutheight, startheight);
+        return false;
+    }
```

Comment:
I think BIP 8 implies that if min_activation_height is set, it must be superior to start_height. Though not explicit "minimum_activation_height should be set to several retarget periods in the future if the startheight is to be very soon after software with parameters is expected to be released."

Reply:
The minimum activation height can be any value, it does not need to be greater than the start_height. If it is less than or equal to the start height, it just means that the locked_in period does not become extended.